src/amoc_api_scenarios_handler.erl

```diff
@@ -10,112 +10,130 @@
+
+-spec init(tuple(), cowboy_req:req(), state()) ->
+    {upgrade, protocol, cowboy_rest}.
 init({tcp, http}, _Req, _Opts) ->
     {upgrade, protocol, cowboy_rest}.
 
--spec rest_init(cowboy:req(), [atom()]) -> {ok, cowboy:req(), state()}.
-rest_init(Req, [Action]) ->
-    {ok, Req, #state{action = Action}}.
+-spec rest_init(cowboy_req:req(), [atom()]) -> {ok, cowboy_req:req(), state()}.
+rest_init(Req, _Opts) ->
+    {ok, Req, #state{}}.
 
--spec allowed_methods(cowboy:req(), state()) -> 
-        {[binary()], cowboy:req(), state()}.
+-spec allowed_methods(cowboy_req:req(), state()) -> 
+    {[binary()], cowboy_req:req(), state()}.
 allowed_methods(Req, State) ->
     {[<<"POST">>, <<"GET">>], Req, State}.
 
--spec content_types_provided(cowboy:req(), state()) -> 
-        {[tuple()], cowboy:req(), state()}.
+-spec content_types_provided(cowboy_req:req(), state()) -> 
+    {[tuple()], cowboy_req:req(), state()}.
 content_types_provided(Req, State) ->
-    {[{<<"application/json">>, process_request}], Req, State}.
+    {[{<<"application/json">>, to_json}], Req, State}.
 
--spec content_types_accepted(cowboy:req(), state()) -> 
-        {[tuple()], cowboy:req(), state()}.
+-spec content_types_accepted(cowboy_req:req(), state()) -> 
+    {[tuple()], cowboy_req:req(), state()}.
 content_types_accepted(Req, State) ->
-    {[{<<"application/json">>, process_request}], Req, State}.
+    {[{<<"application/json">>, from_json}], Req, State}.
+
+%% Request processing functions
+
+-spec to_json(cowboy_req:req(), state()) -> 
+    {jsx:json_text(), cowboy_req:req(), state()}.
+to_json(Req0, State = #state{}) ->
+    {ok, Filenames} = file:list_dir("scenarios"),
+    Filenames2 =
+        lists:filter(
+          fun(X) -> string:right(X, 3) == "erl" end,
+          Filenames),
+
+    Scenarios = 
+    [ erlang:list_to_binary(Y) ||  X <- Filenames2,
+                                   Y <- string:tokens(X, "."),
+                                   Y =/= "erl" ],
+    Reply = jsx:encode([{scenarios, Scenarios}]),
+    {Reply, Req0, State}.
+
+
+-spec from_json(cowboy_req:req(), state()) ->
+    {boolean(), cowboy_req:req(), state()}.
+from_json(Req, #state{}) ->
+    case get_vars_from_body(Req) of
+        {ok, State, Req2} ->
+            ModuleName = State#state.module_name,
+            ModuleSource = State#state.module_source,
+            ScenarioPath = "scenarios/" ++ erlang:binary_to_list(ModuleName),
+            file:write_file(
+              ScenarioPath ++ ".erl",
+              ModuleSource,
+              [write]
+             ),
+            Result = compile_and_load_scenario(
+                       ModuleName,
+                       ScenarioPath),
+            Reply = jsx:encode([{compile, Result}]),
+            Req3 = cowboy_req:set_resp_body(Reply, Req2),
+            {true, Req3, State};
+        {error, bad_request, Req2} ->
+            Reply = jsx:encode([{error, bad_request}]),
```

Comment:
Putting a `bad_request` into the response doesn't look good. Please check it the cowboy isn't sending appropriate error here on tis own. We could put `missing_SOMETHING` here but we don't want to say twice `bad_request`.